xwiki-platform-core/xwiki-platform-livedata/xwiki-platform-livedata-webjar/src/main/webjar/displayers/defaultDisplayer.js

```diff
@@ -0,0 +1,144 @@
+   * @param {Object} logic The logic object associated to the livedata
+   */
+  var Displayer = function (propertyId, entryData, logic) {
+
+    this.logic = logic;
+    this.propertyId = propertyId;
+    this.entryData = entryData;
+
+    this.element = document.createElement("div");
+    this.element.className = "livedata-displayer";
+
+    this.view();
+
+  };
+
+
+  /**
+   * Return the object of parameters to be passed to the viewer and editor functions
+   * @returns {Object} an object containing useful data for displayer: {
+   *  value: the entry property value to be displayed
+   *  property: the property descriptor object
+   *  entry: the entry data object
+   *  config: the configuration object of the displayer, found in the propertyDescriptor
+   *  data: the livedata data object
+   *  logic: the logic instance
+   * }
+   */
+  Displayer.prototype._createParameters = function () {
+    var propertyDescriptor = this.logic.getPropertyDescriptor(this.propertyId);
+    // return the param object
+    return {
+      value: this.entryData[this.propertyId],
+      property: propertyDescriptor,
+      entry: this.entryData,
+      config: propertyDescriptor.displayer || {},
+      data: this.logic.data,
+      logic: this.logic,
+    };
+  };
+
+
+
+  /**
+   * Create viewer element for the displayer
+   * This method can be overriden by other displayers that inherit from this one
+   * Parameters are given by the Displayer.prototype.view function
+   * Must resolve the given promise at the end
+   * @param {Object} defer A jquery promise that must be resolved when the viewer has been created
+   * @param {object} params An object containing useful data for the displayer.
+   *  Param object detail can be found in the Displayer.prototype._createParameters method
+   */
+  Displayer.prototype.createView = function (defer, params) {
+    var element = document.createElement("div");
+    if (params.value !== undefined && params.value !== null) {
+      element.innerText = params.value;
+    }
+    defer.resolve(element);
+  };
+
+
+
+  /**
+   * Create editor element for the displayer
+   * This method can be overriden by other displayers that inherit from this one
+   * Parameters are given by the Displayer.prototype.edit function
+   * Must resolve the given promise at the end
+   * @param {Object} defer A jquery promise that must be resolved when the editor has been created
+   * @param {object} params An object containing useful data for the displayer.
+   *  Param object detail can be found in the Displayer.prototype._createParameters method
+   */
+  Displayer.prototype.createEdit = function (defer, params) {
+
+  };
+
+
+  /**
+   * Call this.createView and append viewer to the displayer root element
+   */
+  Displayer.prototype.view = function () {
+    var self = this;
+    var defer = $.Deferred();
+    var params = this._createParameters();
+
+    this.createView(defer, params);
+    defer.done(function (element) {
```

Comment:
I find it more natural to write:

```
this.createView(params).done(function (element) {
  ...
});
```

Reply:
That was to be sure that the displayer does not forget to resolves the promise, and also to lighten the code in the `createView` methods, as it does not have to explicitly create the promise.
We can create the promise in each displayer `createView` methods implementation instead of receiving it as an argument if you prefer though.

Reply:
I believe it makes the API nicer to use because it allows you to connect promises easily.